galasa-parent/dev.galasa.framework.api.authentication/src/test/java/dev/galasa/framework/api/authentication/OidcProviderTest.java

```diff
@@ -134,63 +167,102 @@ public void testGetJsonWebKeysFromIssuerWithNoKeysReturnsNull() throws Exception
 
-        HttpResponse<Object> mockResponse = new MockHttpResponse<Object>("{" +
-            "\"keys\": [" +
-            "{\"kid\": \"thisisakey\"}," +
-            "{\"kid\": \"thisisanotherkey\"}," +
-            "{\"kid\": \""+targetKeyId+"\"}" +
-            "]}"
-        );
+        JsonObject mockJwks = new JsonObject();
+        JsonArray keysArray = new JsonArray();
+
+        keysArray.add(createMockJwkObject("thisisakey"));
+        keysArray.add(createMockJwkObject("thisisanotherkey"));
+        keysArray.add(createMockJwkObject(targetKeyId));
+        mockJwks.add("keys", keysArray);
+
+        HttpResponse<Object> mockResponse = new MockHttpResponse<Object>(gson.toJson(mockJwks));
 
         HttpClient mockHttpClient = mock(HttpClient.class);
-        when(mockHttpClient.send(any(), any())).thenThrow(new IOException());
+        when(mockHttpClient.send(any(), any())).thenReturn(createMockOidcDiscoveryResponse());
 
         OidcProvider oidcProvider = new OidcProvider("http://dummy-issuer", mockHttpClient);
 
         reset(mockHttpClient);
         when(mockHttpClient.send(any(), any())).thenReturn(mockResponse);
 
         // When...
-        JsonObject key = oidcProvider.getJsonWebKeyFromIssuerByKeyId(targetKeyId, oidcProvider.getIssuer());
+        oidcProvider.getJsonWebKeyByKeyId(targetKeyId);
+
+        // Other calls shouldn't send a request to the issuer as long as the refresh interval hasn't elapsed
+        oidcProvider.getJsonWebKeyByKeyId(targetKeyId);
+        oidcProvider.getJsonWebKeyByKeyId(targetKeyId);
+
+        // Then...
+        // Check that the HTTP client only sent one request
+        verify(mockHttpClient, times(1)).send(any(), any());
+    }
+
+    @Test
+    public void testGetJsonWebKeyByKeyIdWithCachedKeysReturnsCachedKey() throws Exception {
+        // Given...
+        String targetKeyId = "iwantthiskey";
+
+        JsonObject mockJwks = new JsonObject();
+        JsonArray keysArray = new JsonArray();
+
+        keysArray.add(createMockJwkObject("thisisakey"));
+        keysArray.add(createMockJwkObject("thisisanotherkey"));
+        keysArray.add(createMockJwkObject(targetKeyId));
+        mockJwks.add("keys", keysArray);
+
+        HttpResponse<Object> mockResponse = new MockHttpResponse<Object>(gson.toJson(mockJwks));
+
+        HttpClient mockHttpClient = mock(HttpClient.class);
+        when(mockHttpClient.send(any(), any())).thenReturn(createMockOidcDiscoveryResponse());
+
+        OidcProvider oidcProvider = new OidcProvider("http://dummy-issuer", mockHttpClient);
+
+        reset(mockHttpClient);
+        when(mockHttpClient.send(any(), any())).thenReturn(mockResponse);
+
+        // When...
+        JsonWebKey key = oidcProvider.getJsonWebKeyByKeyId(targetKeyId);
 
         // Then...
         assertThat(key).isNotNull();
-        assertThat(key.get("kid").getAsString()).isEqualTo(targetKeyId);
+        assertThat(key.getKeyId()).isEqualTo(targetKeyId);
     }
 
```

Comment:
I was expecting to see a unit test which bumps up the time to force a cache-refresh, and checks that the new set of keys were recovered. Worth doing ?